File: hello.cpp
/** \file  /usr/share/asterisk/agi-bin/hello.cpp
*    Description:  demo code for a call handling prg for asterisk basic agi functions
the  prg sends "order to asterisk by writting AGI COMMANDS STREAM FILE, SEND IMAGE.... on the std ouput 
asterisk will then reply on stdin of the prg  if command was succefull or failed see fct checkresult() has to be called after each call to asterisk 
 Usage: 
- Create an AGI  in /var/lib/asterisk/agi-bin, i.e.: compile with g++ hello.cpp  -o /var/lib/asterisk/agi-bin/testagi 
or in: /usr/share/asterisk/agi-bin
- set permission: chown asterisk:asterisk /var/lib/asterisk/agi-bin/testagi
- Call using EAGI from your dialplan: exten => 100,1,EAGI(restagi) 

 *\version  1.0
 *\date 01/12/12 10:06:50
 *       Revision:  none
 *       Compiler:  gcc
 *
 *\author:  Lyase Damasked
 *        Company:
 *
 * =====================================================================================
 */
#include <stdio.h>
#include <cstdlib>
#include <string.h>
#include "test.h"
using namespace std;
struct agi_env agi;
int tests = 0,
	pass = 0,
	fail =0;

int main () {
int resultcode;
init_agi(); // set up the agi variable and clear stdin of the setup
dump_agi();
// command to play the beep file to current call
fprintf( stderr, "1. Testing 'sendfile'...\n");
printf("STREAM FILE beep \"\"\n");
fflush(stdout);
resultcode = checkresult();
// command to send text msg to current call
fprintf( stderr, "2. Testing 'sendtext'...\n");
printf("SEND TEXT \"hello world\"\n");
fflush(stdout);
resultcode = checkresult();
// command to send an image  to current call
fprintf( stderr, "3. Testing 'sendimage'...\n");
printf("SEND IMAGE asterisk-image\n");
fflush(stdout);
resultcode = checkresult();
// command to read number to  current call
fprintf( stderr, "4. Testing 'saynumber'...\n");
printf("SAY NUMBER 192837465 \"\"\n");
fflush(stdout);
resultcode = checkresult();
// command to read user input will return ascii code of first digit enteredby  current call
fprintf( stderr, "5. Testing 'waitdtmf'...\n");
printf("WAIT FOR DIGIT 1000\n");
fflush(stdout);
resultcode = checkresult();
// command to record sound message of 30s or until user hits keyboard in  current call
//default location for file is usr/share/asterisk/sounds/testagi.gsm
// [Dec  2 06:18:23] WARNING[12108]: file.c:1160 ast_writefile: Unable to open file /usr/share/asterisk/sounds/testagi.gsm: Permission denied

fprintf( stderr, "6. Testing 'record'...\n");
printf("RECORD FILE /tmp/testagi gsm 1234 3000\n");
fflush(stdout);
resultcode = checkresult();
// command to playback sound file to  current call
fprintf( stderr, "6a. Testing 'record' playback...\n");
printf("STREAM FILE /tmp/testagi \"\"\n");
fflush(stdout);
resultcode = checkresult();
fprintf(stderr,"================== Complete ======================\n");
fprintf(stderr,"%d tests completed, %d passed, %d failed\n",tests,pass,fail);
fprintf(stderr,"==================================================\n");
}

void dump_agi(){
fprintf(stderr,"agi_request: %s\n",agi.request);
fprintf(stderr,"agi_channel: %s\n",agi.channel);
fprintf(stderr,"agi_lang: %s\n",agi.lang);
fprintf(stderr,"agi_type: %s\n",agi.type);
fprintf(stderr,"agi_callerid: %s\n",agi.callerid);
fprintf(stderr,"agi_dnid: %s\n",agi.dnid);
fprintf(stderr,"agi_context: %s\n",agi.context);
fprintf(stderr,"agi_extension: %s\n",agi.extension);
fprintf(stderr,"agi_priority: %d\n",agi.priority);
}

void init_agi(){
char line[200]; // big just incase
char *name,*value,*temp;
while(fgets(line,200,stdin)){
if(!strncmp(line,"agi_",4)){
	// lets parse this line
	if(name = strtok(line,":")){
		value = strtok(NULL,"\n"); // grab remaining minus the newline
		value++; // an attempt to get rid of the leading space. 
		if(!strncmp(name,"agi_request",11)){
			strncpy(agi.request,value,100);
			}
		if(!strncmp(name,"agi_channel",11)){
			strncpy(agi.channel,value,80);
			}
		if(!strncmp(name,"agi_language",12)){
			strncpy(agi.lang,value,20);
			}
		if(!strncmp(name,"agi_type",8)){
			strncpy(agi.type,value,100);
			}
		if(!strncmp(name,"agi_callerid",12)){
			strncpy(agi.callerid,value,100);
			}
		if(!strncmp(name,"agi_dnid",8)){
			strncpy(agi.dnid,value,100);
			}
		if(!strncmp(name,"agi_context",8)){
			strncpy(agi.context,value,80);
			}
		if(!strncmp(name,"agi_extension",8)){
			strncpy(agi.extension,value,80);
			}
		if(!strncmp(name,"agi_priority",8)){
			agi.priority = atoi(value);
			}
		}
	}else{
		break;
		}
}
}
						
int checkresult (){
char *parts;
char result[200]; 
int code;
if(!fgets(result,200,stdin)){
	exit(0); // should only get here on error or EOF we could not read in stdin a reply from asterisk at least 200char
	}
tests++; // we got a message back a test was run
if(strncmp(result,"200",3)){ 
	fail++;
	return 0;
	}else{
		pass++;
		if(parts = strtok(result,"=")){
			if(parts = strtok(NULL,"\n")){ // grab remaining minus the newline
				code = atoi(parts);
				}
			else{
				exit(0); // means we didn't get a result code
				}
			}else{
				exit(0); // means strtok failed originally
				}
			}
	return code;
}

File: hello.pl
#!/usr/bin/perl
# this script is called when asterisk gets incoming call 
# you are "printing to stdouput the commands to asterisk to manage user
#ex: play file 
#get dtmf
#...
use strict;
use POSIX;

$|=1;
# Set up some variables
my %AGI; my $tests = 0; my $fail = 0; my $pass = 0;
while(<STDIN>) {
        chomp;
        last unless length($_);
        if (/^agi_(\w+)\:\s+(.*)$/) {
                $AGI{$1} = $2;
        }
}
print STDERR "AGI Environment Dump:\n";
foreach my $i (sort keys %AGI) {
        print STDERR " -- $i = $AGI{$i}\n";
}
sub checkresult {
        my ($res) = @_;
        my $retval;
        $tests++;
        chomp $res;
        if ($res =~ /^200/) {
                $res =~ /result=(-?\d+)/;
                if (!length($1)) {
                        print STDERR "FAIL ($res)\n";
                        $fail++;
                } else {
                        print STDERR "PASS ($1)\n";
                        $pass++;
                }
        } else {
                print STDERR "FAIL (unexpected result '$res')\n";
                $fail++;
        }
}

					print STDERR "1.  Testing 'streaming  sound file and tones'...";
print "STREAM FILE beep \"\"\n";
#tt-weasels.gsm is in sftp://root@debian-asterisk/usr/share/asterisk/sounds/en_US_f_Allison/
print "STREAM FILE tt-weasels \"\"\n";
# the sftp://root@debian-asterisk/usr/share/asterisk/sounds/outfile.gsmhas poor quality  is created with "sox infile.wav outfile.gsm"
#print "STREAM FILE outfile \"\"\n";
my $result = <STDIN>;
&checkresult($result);
#does not work 
#file does not exist print "STREAM FILE 8 \"\"\n";
#file not found print "STREAM FILE testagi \"\"\n";
 
#my $result = <STDIN>;
#&checkresult($result);

					print STDERR "2.  Testing 'sendtext'...";
print "SEND TEXT \"hello world\"\n";
my $result = <STDIN>;
&checkresult($result);

					print STDERR "3.  Testing 'sendimage'...";
print "SEND IMAGE asterisk-image\n";
my $result = <STDIN>;
&checkresult($result);
					print STDERR "4.  Testing 'saynumber'...";
print "SAY NUMBER 192837465 \"\"\n";
my $result = <STDIN>;
&checkresult($result);
print "STREAM FILE beep \"\"\n";
my $result = <STDIN>;
&checkresult($result);
					print STDERR "5.  Read dtmf on incoming line 'waitdtmf  10s'...\n";
# This application returns '200 result=49' the result is decimal ASCII value of the digit that was pressed.
# ascii 1=>49
print "WAIT FOR DIGIT  10000 \n";

my $result = <STDIN>;
&checkresult($result);
#;my $res =($result=~/result=\(-?\d+)/g);
my $res=$result =~ /=(\d+)/;

if(($result=~m/result=(\d+)/)){
#correction for ascii
$res=$1 -48;
}


my $user = `whoami`;
my $now = strftime("%Y.%m.%d.%R", localtime(time));
my $file = "/tmp/hello.pl_AGI_script_debug_output_${user}_${now}.txt";

open FILE, ">$file" or die "unable to open $file $!";

print FILE "this line is fake to print a first line  to your $file output file if you don't see entire output this script may have crashed in asterisk check your sintax \n";
print FILE "complete reply from asterisk: $result\n";
print FILE " extracted  result  res is :  $res\n";
					print STDERR "5b.  got escape digit'...\n";
					print FILE "5b.  got escape digit'...\n";
print "STREAM FILE beep \"\"\n";
my $result = <STDIN>;
&checkresult($result);
					print STDERR "6.  Testing 'record until digit 1 or 10s'...";
					 print FILE  "6.  Testing 'record until digit 1 or 10s'...";
print "RECORD FILE testaginewlocation gsm 1 10000";
my $result = <STDIN>;
&checkresult($result);
print "STREAM FILE beep \"\"\n";
my $result = <STDIN>;
&checkresult($result);
					print STDERR "6a.  Testing 'record playback'...";
					print FILE  "6a.  Testing 'record playback'...";
print "STREAM FILE testaginewlocation \"\"\n";
my $result = <STDIN>;
&checkresult($result);
					print STDERR "7a.  Testing 'hangup'...";
					print FILE  "7a.  Testing 'hangup'...";
print "HANGUP \"\"\n";
my $result = <STDIN>;
&checkresult($result);
print STDERR "================== Complete ======================\n";
print STDERR "$tests tests completed, $pass passed, $fail failed\n";
print STDERR "==================================================\n";
close (FILE); 
exit(0);
